src/Microsoft.AspNetCore.Http.Abstractions/Routing/RouteValueDictionary.cs

```diff
@@ -0,0 +1,614 @@
+                    Add(kvp.Key, kvp.Value);
+                }
+
+                return;
+            }
+
+            if (values is IEnumerable<KeyValuePair<string, string>> stringValueEnumerable)
+            {
+                foreach (var kvp in stringValueEnumerable)
+                {
+                    Add(kvp.Key, kvp.Value);
+                }
+
+                return;
+            }
+
+            if (values != null)
+            {
+                var storage = new PropertyStorage(values);
+                _propertyStorage = storage;
+                _count = storage.Properties.Length;
+                return;
+            }
+        }
+
+        /// <inheritdoc />
+        public object this[string key]
+        {
+            get
+            {
+                if (string.IsNullOrEmpty(key))
+                {
+                    throw new ArgumentNullException(nameof(key));
+                }
+
+                object value;
+                TryGetValue(key, out value);
+                return value;
+            }
+
+            set
+            {
+                if (string.IsNullOrEmpty(key))
+                {
+                    throw new ArgumentNullException(nameof(key));
+                }
+
+                // We're calling this here for the side-effect of converting from properties
+                // to array. We need to create the array even if we just set an existing value since
+                // property storage is immutable. 
+                EnsureCapacity(_count);
+
+                var index = FindInArray(key);
+                if (index < 0)
+                {
+                    EnsureCapacity(_count + 1);
+                    _arrayStorage[_count++] = new KeyValuePair<string, object>(key, value);
+                }
+                else
+                {
+                    _arrayStorage[index] = new KeyValuePair<string, object>(key, value);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Gets the comparer for this dictionary.
+        /// </summary>
+        /// <remarks>
+        /// This will always be a reference to <see cref="StringComparer.OrdinalIgnoreCase"/>
+        /// </remarks>
+        public IEqualityComparer<string> Comparer => StringComparer.OrdinalIgnoreCase;
+
+        /// <inheritdoc />
+        public int Count => _count;
+
+        /// <inheritdoc />
+        bool ICollection<KeyValuePair<string, object>>.IsReadOnly => false;
+
+        /// <inheritdoc />
+        public ICollection<string> Keys
+        {
+            get
+            {
+                EnsureCapacity(_count);
+
+                var array = _arrayStorage;
+                var keys = new string[_count];
+                for (var i = 0; i < keys.Length; i++)
+                {
+                    keys[i] = array[i].Key;
+                }
+
+                return keys;
+            }
+        }
+
+        IEnumerable<string> IReadOnlyDictionary<string, object>.Keys
```

Comment:
`IEnumerable<string> IReadOnlyDictionary<string, object>.Keys => Keys` for 💯 points

Reply:
https://github.com/aspnet/Routing/pull/708